templates/fine-tune-stable-diffusion/README.ipynb

```diff
@@ -0,0 +1,390 @@
+    "    \"--resolution=1024\",\n",
+    "    \"--train_batch_size=1\",\n",
+    "    \"--gradient_accumulation_steps=1\",\n",
+    "    \"--learning_rate=1e-4\",\n",
+    "    \"--lr_scheduler=constant\",\n",
+    "    \"--lr_warmup_steps=0\",\n",
+    "    \"--max_train_steps=100\",\n",
+    "    \"--checkpointing_steps=100\",\n",
+    "    # Use the first prompt as a sample to generate during training.\n",
+    "    f\"--validation_prompt={PROMPTS[0]}\",\n",
+    "    \"--validation_epochs=25\",\n",
+    "    \"--seed=0\",\n",
+    "] + ([\"--report_to=wandb\"] if os.environ.get(\"WANDB_API_KEY\") else [])\n",
+    "\n",
+    "TRAINING_ARGS = parse_args(input_args=cmd_line_args)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Launch distributed training with Ray Train\n",
+    "\n",
+    "To run distributed training, we'll use a `ray.train.torch.TorchTrainer` to request GPU workers and connect them together in a distributed worker group. Then, when the workers run the training script, HuggingFace Accelerate detects this distributed process group and sets up the model to do data parallel training.\n",
+    "\n",
+    "A few notes:\n",
+    "* `ray.init(runtime_env={\"env_vars\": ...})` sets the environment variables on all workers in the cluster -- setting the environment variable in this notebook on the head node is not enough in a distributed setting.\n",
+    "* `train_fn_per_worker` is the function that will run on all distributed training workers. In this case, it's just a light wrapper on top of the `diffusers` example script that copies the latest checkpoint to shared cluster storage.\n",
+    "* `ScalingConfig` is the configuration that determines how many workers and what kind of accelerator to use for training. Once the training is launched, **Anyscale will automatically scale up nodes to meet this resource request!**\n",
+    "\n",
+    "The result of this fine-tuning will be a fine-tuned LoRA model checkpoint at `MODEL_CHECKPOINT_PATH`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_CHECKPOINT_PATH = os.environ[\"ANYSCALE_ARTIFACT_STORAGE\"] + \"/checkpoint-final\"\n",
+    "\n",
+    "print(\"Final checkpoint will be uploaded to: \", MODEL_CHECKPOINT_PATH)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "import ray\n",
+    "import ray.train\n",
+    "from ray.train.torch import TorchTrainer\n",
+    "\n",
+    "from train_dreambooth_lora_sdxl import main\n",
+    "from utils import download_from_cloud, upload_to_cloud\n",
+    "\n",
+    "\n",
+    "# Set environment variables across the entire cluster.\n",
+    "ray.init(\n",
+    "    runtime_env={\n",
+    "        \"env_vars\": {\"WANDB_API_KEY\": os.environ.get(\"WANDB_API_KEY\")}\n",
```

Comment:
This actually crashes if you don't have WANDB_API_KEY set. Make it optional / comment out this block by default?